geode-core/src/main/java/org/apache/geode/distributed/internal/InternalDistributedSystem.java

```diff
@@ -853,12 +876,35 @@ private void startSampler() {
   }
 
   /**
+   * Starts a locator in this JVM iff the distribution config wants one started.
+   *
+   * @return the membershipLocatorArg if the distribution config has no locator specified;
+   *         otherwise starts a new InternalLocator and returns its associated MembershipLocator
+   *
    * @since GemFire 5.7
+   * @param membershipLocatorArg an already-running MembershipLocator, or null if none is known
    */
-  private void startInitLocator() throws InterruptedException {
-    String locatorString = originalConfig.getStartLocator();
-    if (locatorString.length() == 0) {
-      return;
+  private MembershipLocator startInitLocator(
+      final MembershipLocator<InternalDistributedMember> membershipLocatorArg)
+      throws InterruptedException {
+
+    final String locatorString = originalConfig.getStartLocator();
+    final boolean shouldStartLocator = locatorString.length() > 0;
+
+    if (!shouldStartLocator) {
+      return membershipLocatorArg;
+    }
+
+    /*
+     * Defensive coding here. InternalLocator's constructor is keeping track of a singleton instance
+     * and that constructor will throw an exception if we try to have more than one
+     * InternalLocator in existence at the same time. So in that sense we shouldn't need this check.
+     * Nevertheless, we'd like the code over here to not rely on that singleton bookkeeping over
+     * there. So if that singleton logic goes away, this code here might save us.
+     */
+    if (membershipLocatorArg != null && shouldStartLocator) {
+      throw new IllegalStateException("Internal error: trying to start a second locator "
+          + "in this JVM");
```

Comment:
Reviewers: I'm looking for your critique of this exception. If somebody passed in a locator, and also specified config parameters to cause this routine to start a locator, it seems to me we want to stop and inform them of the conflict.

Reply:
I'm not sure I would describe this as an "Internal error" since it seems like it results from a user's misconfiguration?

Could we be more specific here - like "Setting the start-locator property is not supported in dedicated locators launched through gfsh or the Java Locator class?"

Reply:
fixed